engine/src/tangl/vm38/traversable.py

```
@@ -0,0 +1,670 @@
+
+        The return edge targets the predecessor (the call site) and starts
+        the pipeline at ``return_phase`` (skipping phases already completed
+        before the call).
+
+        Raises
+        ------
+        ValueError
+            If ``return_phase`` is not set (not a call edge).
+        """
+        if self.return_phase is None:
+            raise ValueError(f"{self!r} is not a call edge (return_phase is None)")
```

Ruff (0.15.0)

[warning] Avoid specifying long messages outside the exception class

(TRY003)